### PR TITLE
Improve throughput of Environment.GetEnvironmentVariables()

### PR DESCRIPTION
**Pull request by stephentoub**
Use IndexOf{Any} to search for positions rather than open-coded loops, taking advantage of vectorization to improve throughput.

```C#
[Benchmark]
public IDictionary Get() => Environment.GetEnvironmentVariables();
```

With the environment variables in my command prompt:

| Method |        Job |           Toolchain |     Mean |    Error |   StdDev |   Median | Ratio |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|------- |----------- |-------------------- |---------:|---------:|---------:|---------:|------:|-------:|-------:|------:|----------:|
|    Get | Job-PPEVTD | \master\corerun.exe | 35.17 us | 0.423 us | 0.396 us | 35.10 us |  1.00 | 5.1270 | 0.8545 |     - |  31.43 KB |
|    Get | Job-PCZKGH |     \pr\corerun.exe | 14.61 us | 0.278 us | 0.726 us | 14.33 us |  0.41 | 5.1270 | 0.8545 |     - |  31.43 KB |

Contributes to https://github.com/dotnet/runtime/issues/44598

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs [hunk 1]
```diff
@@ -92,44 +79,43 @@ public static unsafe IDictionary GetEnvironmentVariables()
                 // environment block (as =C:=pwd) and the program's exit code is
                 // as well (=ExitCode=00000000).
 
+                // Search for terminating \0\0 (two unicode \0's).
+                // IndexOf processes memory in aligned chunks, and thus it won't crash even if it accesses memory beyond the null terminator.
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs:83]: IndexOf searches for the first char (ie \0 in this case) and then compares the rest.

It should better to just search for the next terminating zero inside the loop, using `wcslen`

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs:83]: Sounds good.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs:83]: Just to clarify,  did you mean:
a) replace the IndexOf with my own loop that uses wcslen + checking the next character,  or
b) delete the IndexOf, and change the main loop to use wcslen on each iteration to find the next terminator
?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs:83]: Yep (b)

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs]: That cleans things up nicely, and is a tad faster.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs [hunk 2]
```diff
@@ -58,93 +58,68 @@ private static void SetEnvironmentVariableCore(string variable, string? value)
...
-                        while (block[i] != 0)
+                        // Add the key and value.
+                        string key = new string(variable.Slice(0, i));
+                        string value = new string(variable.Slice(i + 1));
+                        try
                         {
-                            i++;
+                            // Add may throw if the environment block was corrupted leading to duplicate entries.
+                            // We allow such throws and eat them (rather than proactively checking for duplication)
+                            // to as to provide a non-fatal notification about the corruption.
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs:107]:
```suggestion
                            // to provide a non-fatal notification about the corruption.
```

**Review by jkotas — APPROVED** [review]
Nice!

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetEnvironmentStrings.cs [hunk 3]
```diff
@@ -7,7 +7,7 @@ internal static partial class Interop
 {
     internal static partial class Kernel32
     {
-        [DllImport(Libraries.Kernel32, EntryPoint = "GetEnvironmentStringsW", SetLastError = true, CharSet = CharSet.Unicode, BestFitMapping = false, ExactSpelling = true)]
-        internal static extern unsafe char* GetEnvironmentStrings();
+        [DllImport(Libraries.Kernel32, ExactSpelling = true)]
+        internal static extern unsafe char* GetEnvironmentStringsW();
```
Comment (jkotas) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetEnvironmentStrings.cs:11]: Has to be also fixed up in https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/ecalllist.h#L1055

Comment (stephentoub) [reply src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetEnvironmentStrings.cs:11]: Ah, the pal, thanks.

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.FreeEnvironmentStrings.cs [hunk 4]
```diff
@@ -7,7 +7,7 @@ internal static partial class Interop
 {
     internal static partial class Kernel32
     {
-        [DllImport(Libraries.Kernel32, EntryPoint = "FreeEnvironmentStringsW", SetLastError = true, CharSet = CharSet.Unicode, BestFitMapping = false, ExactSpelling = true)]
-        internal static extern unsafe bool FreeEnvironmentStrings(char* lpszEnvironmentBlock);
+        [DllImport(Libraries.Kernel32, ExactSpelling = true)]
+        internal static extern unsafe int FreeEnvironmentStringsW(char* lpszEnvironmentBlock);
```
Comment (jkotas) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.FreeEnvironmentStrings.cs:11]:
```suggestion
        internal static extern unsafe BOOL FreeEnvironmentStringsW(char* lpszEnvironmentBlock);
```

**Comment by adamsitnik** [comment]
@stephentoub do you have any plans on improving the Unix implementation as well? This could improve the developer loop experience on macOS

**Comment by am11** [comment]
@adamsitnik, `Environment.Variables.Windows.cs` is compiled for both Windows and Unix under feature coreclr.

**Comment by adamsitnik** [comment]
> is compiled for both Windows and Unix under feature coreclr

@am11 it might be, but according to the data shared in https://github.com/dotnet/runtime/issues/41871, getting a single env var on macOS is still 4 times slower compared to Windows (using same hardware). And the time is not spent in the method that @stephentoub has optimized in this particular PR (see https://github.com/dotnet/runtime/issues/866#issuecomment-547807782 for details)

**Comment by stephentoub** [comment]
> do you have any plans on improving the Unix implementation as well? This could improve the developer loop experience on macOS

As @am11 said, this is the implementation of GetEnvironmentVariables for both Windows and Unix.  GetEnvironmentVariable (singular) doesn't use this, on any platform.

**Comment by adamsitnik** [comment]
Are there any plans to improve the performance of Unix implementation of the GetEnvironmentVariable (singular) method?

**Comment by stephentoub** [comment]
The costs you highlight for GetEnvironmentVariable are because the runtime reads and stores the UTF8 env vars on first use,  and then calls to GetEnvironmentVariable need to transcode the UTF16 key being searched for, which is done linearly.  To make GetEnvironmentVariable faster in microbenchmarks, the runtime could instead pre-transcode them all and potentially store them in a hashtable. The problem, though, is trying to optimize for microbenchmark on the throughput of GetEnvironmentVariable will likely lead you to do the "wrong" thing.  No one should be calling GetEnvironmentVariable on a hot path; instead,  it's generally used once for a given key, often during startup / first use of some code that checks it for configuration info and then never looks at it again. We could spend way more time transcoding all env vars, find we made a benchmark.net test of GetEnvironmentVariable much faster, but actually slowed down the overall startup of an app with no measurable reward to show for it. If you'd like to prototype that and can demonstrate it does help with the overall startup of various app types (e.g. because it turns out we end up looking for lots of environment variables, maybe multiple times), then removing could certainly consider that direction.

**Comment by kunalspathak** [comment]
Perf lab results shows improvements - https://github.com/DrewScoggins/performance-2/issues/3515

**Pull request merged**